### ISSUE TITLE
🌟Feature request: Focus to first input on click when Pin Input is fully empty regardless of which individual input was clicked

### ISSUE THREAD

**Issue by yekta** [melt-ui/melt-ui#488]
**Is your feature request related to a problem? Please describe.**
Kinda. When the `Pin Input` is fully empty and when I click on any of the individual inputs, I expect the first input field to be focused since there is not much point for a user to want to focus to say 4th input if the pin input field is fully empty.

**Describe the solution you'd like**
If `Pin Input` is fully empty, I would like the first input field to be focused when I click anywhere in the `Pin Input`. This makes hitting the `Pin Input` target easier. I don't see a use case for actually wanting to click say 4th input while the entire `Pin Input` field is empty. Usually you just want to click the field as fast as possible and enter the pin as fast as possible.

**Comment by yekta** [comment]
> This is tricky. Here are my thoughts.
> 
> This would make it similar in behaviour to a normal input would, with just different looks. If that is desired, what you describe is ideal. But it goes beyond that.
> 
> Let's imagine we have a pin input with 4 characters.
> 
> * Fully empty and I click? Focus on the first.
> * If we have `1 2 ○ ○`, when I click on an empty field, it should probably always focus the first empty field.
> * What if I have `1 2 3 4`, and delete the first two numbers? Should it become `3 4 ○ ○`, like a normal input would, or `○ ○ 3 4`, how we visually would expect it to?
> * If I click on an empty field in `○ 2 ○ 4`, what should it do?
> * ...
> 
> I think this will lead us to a road where the behaviour of pin-input becomes unpredictable. There are some things we can do to improve experience though.
> 
> * Make it so only the first empty element has tabIndex="0". If all are filled, make the last filled item have tabIndex="0".
> * Add Home and End key handlers
> 
> And one thing that we already do: No matter which input is focused when you paste a code. it will do its best to fill the pin input accordingly. e.g. if the code is `1234`, and I'm focused on the 3rd input and I press paste, the result will be `1 2 3 4`, and not `○ ○ 1 2`.

I don’t think you need to answer all those questions to implement the rule I suggested 🤔

Rest of the rules you can evaluate completely separately. I’m suggesting a single rule, based on a single condition. I don’t think there is a single use-case for wanting to click to say the third field if all fields are empty. If anything, I think the rule I suggested makes the pin field more predictable not less. I have opinions about other rules as well but I didn’t write them in detail because majority of the gain in my opinion comes from the rule I suggested. Rest of the rules include more complicated conditions.

**Comment by TGlide** [comment]
> I don’t think you need to answer all those questions to implement the rule I suggested 🤔

But we do! Because naturally when we implement something like this, these questions appear:
`"Why does this behaviour only happen when its all empty? Shouldn't the first **empty** input be selected always?"`

And then comes the other questions I mentioned. If we view all behaviours in isolation, clashes begin to appear. These things need to be thought out in the context they're in, not separately.

For the sake of being comprehensive though, there are arguments that could be made against it, in isolation. Some may find it unexpected for the focus to move out the thing you clicked, and find it janky. It's also non-standard when comparing with other pin-inputs. 

I personally think the bigger problems arise when looking it from a wider context though.

And...

> it being very easy to reason about and implement (I did it myself

This is the beauty of it. We offer plenty of ways to customize **everything** in Melt. I'd be up to having an example in the docs where this is shown, to be honest. I don't think it should be a default though.

**Comment by yekta** [comment]
> > I don’t think you need to answer all those questions to implement the rule I suggested 🤔
> 
> But we do! Because naturally when we implement something like this, these questions appear: `"Why does this behaviour only happen when its all empty? Shouldn't the first **empty** input be selected always?"`
> 
> And then comes the other questions I mentioned. If we view all behaviours in isolation, clashes begin to appear. These things need to be thought out in the context they're in, not separately.
> 
> For the sake of being comprehensive though, there are arguments that could be made against it, in isolation. Some may find it unexpected for the focus to move out the thing you clicked, and find it janky. It's also non-standard when comparing with other pin-inputs.
> 
> I personally think the bigger problems arise when looking it from a wider context though.
> 
> And...
> 
> > it being very easy to reason about and implement (I did it myself
> 
> This is the beauty of it. We offer plenty of ways to customize **everything** in Melt. I'd be up to having an example in the docs where this is shown, to be honest. I don't think it should be a default though.

Doesn’t need to be a default, but I think most devs will appreciate the option.

From my perspective I have plenty of questions about the pin input, one of which is:

“Why it doesn’t focus to the first input when I click while all fields are empty?”.

Suggestion I made makes this question go away. Then you can have the question of:

“Given that happens, why this other thing didn’t happen”.

But, even if you don’t implement the feature, I will still have that second question, just in this form:

“Why this other thing didn’t happen?”

So to me, answering those other questions are not a prerequisite, they are other nice to haves. But I’m not maintaining the library 🙂

Regarding “out of the thing you clicked”: I think the amount of people that would think that would be dwarfed by the amount of people that would think: “why didn’t it focus to the thing I actually wanted it to focus”.

I can also provide answers to other questions you raised btw, but if you don’t agree with the one that was most clear to me by a pretty large margin, I don’t think me answering those questions would mean much.

**Comment by TGlide** [comment]
> Doesn’t need to be a default, but I think most devs will appreciate the option.

Too many options can make things confusing. If it were only an option to focus the first input, fine, but I don't want to introduce this option in isolation as I don't think it makes sense. And I don't want to introduce it as an option with other things because then, why do we have two ways of doing things? 

This is why I add examples in the docs of how you can introduce alternate behaviours.

> From my perspective I have plenty of questions about the pin input, one of which is:
> 
> “Why it doesn’t focus to the first input when I click while all fields are empty?”.
> 
> Suggestion I made makes this question go away. Then you can have the question of:
> 
> “Given that happens, why this other thing didn’t happen”.
> 
> But, even if you don’t implement the feature, I will still have that second question, just in this form:
> 
> “Why this other thing didn’t happen?”
> 
> So to me, answering those other questions are not a prerequisite, they are other nice to haves. But I’m not maintaining the library 🙂

It doesn't work that way. 

If we want to implement a feature, we need a reason.

Feature: Focus first input of Pin Input when you click it and its empty.
Reason: Because users will most likely want to write it from start to finish

If we follow that, that reason leads us to think about other features that that reason brings about. E.g. If `users will most likely want to write it from start to finish`, then what do we want to implement to aid that? If I don't think about other cases that follow that reasoning, that's bad design on my part. 

Or, another way of thinking is this. I introduced a behaviour in which, sometimes, clicking on an empty input moves focus to other places. But sometimes it doesn't. This brings questions to a curious individual. What are these cases? And why? What's the reasoning behind it? Is it clear, simple, straight-forward? I don't think they are

> Regarding “out of the thing you clicked”: I think the amount of people that would think that would be dwarfed by the amount of people that would think: “why didn’t it focus to the thing I actually wanted it to focus”.

I disagree. You click something, it focuses to the thing you clicked. That is the norm, and so it is expected. You may ask yourself, of course, "wouldn't it be better to do x?". But it wouldn't be unexpected that clicking an input leads the browser to focus the input you clicked.

---

I'll be closing this issue as not-planned, for the reasons mentioned. I have however created a separate issue (#494) to add examples in the docs of how to implement this behaviour, as it's quite painless to do so.

**Comment by yekta** [comment]
Okay then. As an extra note, I think this is pretty similar in thinking and UX to your paste feature. If you consider pin input's individual inputs to have that level of separation (as in they are very individual "things" always when you click them), not sure why pasting on the 3rd field can affect field 1 and 2. Since the "norm" is, when you paste into a thing, you expect it to be pasted into that thing. Yet, pasting into the 3rd field can affect 1 and 2, which could raise similarly many questions to the ones you suggested to my suggestion. But I do like that behaviour, since I don't see anyone wanting to paste the first letter of a 4 letter string into the 3rd field, instead of wanting to paste 4 letters from start to finish to 4 letter input. That suggests me you considered `Pin Input` as a singular thing at least 1 time, without it rising many other questions and preventing you to implement that specific behaviour that improves UX (without answering those other questions).

**Comment by TGlide** [comment]
> Yet, pasting into the 3rd field can affect 1 and 2, which could raise similarly many questions to the ones you suggested to my suggestion

It's much more clearly defined and self-contained. When you paste something, where will the pasted content go? There's only two situations. **If the pasted content is equal to or bigger in length** to the number of inputs, it will start pasting from the first input. **Otherwise**, it will paste normally. 

I'd say there is only one question that could arise from this behaviour, which is, why don't we always paste from the start? To which I say there is a straightforward conclusion, which is, why would you only partially copy a code and paste it? If you do, there's no way to know where the partial content fits in there, so no reason to choose where to paste it besides where the user tried to paste it from.

>  That suggests me that you considered Pin Input as a singular thing at least 1 time, without the need to answer many other questions it rises because it improves UX

I have considered Pin Input as a singular **thing** (rather than a singular `<input />`) many times. When you fill in one of the inputs, it goes to the next. When you press backspace on an empty input, it goes to the previous one. You have kbd navigation between them besides normal tabbing. And I view these behaviours in tandem, to improve the UX.

--- 

There is an alternative I propose, that could be satisfactory and that I'd be glad in implementing, thinking more about it. While the way of inputting currently is more free-form, I do see a use case for a more rigid behaviour.

E.g. Let's say we have `createPinInput({ rigid: true })`
- When you click the pin input, it focus the first empty input
- You may not use arrow keys or mouse to navigate to an input which is not filled, or is not the first empty input
- Deleting a field in the middle of other filled inputs leads to the inputs to the front going back 1 spot. E.g. we have `1 2 3 ○` and delete 2, it will become `1 3 ○ ○`. P.S. You can still focus the second input and write something else to replace it

\* *alternatively* \*

- When you click the pin input, it focus the first empty input
- You may not use arrow keys or mouse to navigate to an input which is not the **last** filled, or is not the first empty input

Would love to hear your thoughts on this.

**Comment by yekta** [comment]
I don't agree paste behaviour being much more clearly defined and self contained (I think the behaviour is actually less clear than the one I suggested, yet I still like it all the same 🙂), but I'm gonna skip the counter argument on that part. Looks like we don't agree on some fundamental definitions there and it will cause an infinite back and forth which I'm not after.

When it comes to the second part:

I don't have a strong preference between the two you approaches you suggested. I like both of them better than the current behaviour. If I have to pick one, I'd pick the second one.

To me, `Pin Input` is mostly for visual separation (on the part of the user of the app, not the library). It's also mostly a visual separation on other pin inputs I've used (mostly on mobile). I don't see a reason for needing to focus on the third input if both second and the third input is empty. It follows a similar logic to me not seeing a reason to focus on the third input when all inputs are empty.

However, when the situation is this: `a b ○ ○`. I do see a reason for someone to want to focus on `a` and wanting to clear `a` and everything to the right of it (it would be uncommon, but not as uncommon as someone wanting to focus on the third input when all are empty). So this would clash with your first rule.

So this then turns into a set of disputable rules (which I personally don't have a problem with) but things will get complicated very quick if you are going for a complete set of rules. That's why I kept the rule I suggested as simple, and as indisputable as possible (under any circumstance). But if you are reaching for complete set of rules, between the things you suggested I'd go for the second one. I don't think it's complete though, and I think it still raises other questions you might want to answer. I think it's better than the current version and that's enough for me.

**Comment by TGlide** [comment]
`Pin Input` acts like a singular input at all times.

While this is not a bad idea per se, it eleminates some enhancements on our end. Such as the paste, and when you're focused on one of the inputs and write a number, normally I'd like to override it instead of adding a number and pushing the others forward.

I'll have to think this one through. In any case, I'll re-open the issue to implement the rigid/strict/enforced/idk mode later!

**Comment by yekta** [comment]
> `Pin Input` acts like a singular input at all times.
> 
> While this is not a bad idea per se, it eliminates some enhancements on our end. Such as the paste, and when you're focused on one of the inputs and input a number, normally I'd like to overwrite it instead of adding a number and pushing the others forward.
> 
> I'll have to think this one through. In any case, I'll re-open the issue to implement the rigid/strict/enforced/idk mode later!

I agree, there is no need to get rid of the nice to use and logical behaviour you already have for paste. I tried to keep the rule set as simple as possible for the sake of consistency and with the intention of showing that if you are going for a disputable set of rules, possibilities are endless :) I personally don't care much about it as long as things individually make sense, and produce "consistent enough" result.

The rule can also be:

- Pin Input acts like a singular input almost always, with the exception of paste.

That would still make sense to me, though would be disputable and partially 'inconsistent' (why does it work like a singular input sometimes and not other times, again not that I personally care, I think it's just fine). Just like your two suggestions make sense, while being disputable. My main request with this is still the same however (because I think it covers "most" of the use cases with a very simple rule I don't think anyone would be against + very little effort): If all inputs are empty and I click the `Pin Input`, it focuses to the first input.

Here is what I think will happen 'almost' all of the time:

1- People will click anywhere on the `Pin Input` and write the code in 1 go. Maybe make 1 mistake, delete the mistake and continue writing.
2- People will come with the intention of pasting the code they just copied. They click anywhere on the `Pin Input` (or the first field) and paste it.

Here is what I think might happen rarely:

1- People will write the code until say the second field, focus somewhere else, and come back to complete it by clicking anywhere on the `Pin Input` (or the first empty field).
2- People will write the code until say the second field, focus somewhere else, then click the first field.

Here is what I think will almost never happen:

1- People will want to focus on the third field non-mistakingly while all fields are empty.

So I guess your rule set depends on how much time you are willing to spend thinking & implementing, how big of a percentage of the use case you want to cover, and how simple/consistent of a rule set you want. Given I don't know your consistency, coverage and worth metrics/threshold, hard to make a good overall suggestion.

Anyways, happy to see that you are willing to consider the behaviour. I think that's enough arguments on my side 😄 I'll be back to see what you've decided.

**Comment by huntabyte** [comment]
Closing as not planned in the immediate future, as there hasn't been enough demand for this from the community. Should demand arise, we can reopen and prioritize this issue.

**Comment by TGlide** [comment]
The next version of Melt will have this feature.
